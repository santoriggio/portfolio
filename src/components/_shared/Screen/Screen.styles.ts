import { ThemedStyle } from "@/theme";
import { ViewStyle } from "react-native";
import {
  ContentStyleProps,
  HeaderLargeTitleStyleProps,
  HeaderStyleProps,
  HeaderTitleStyleProps,
} from "./Screen.types";

export const $basePageStyle: ViewStyle = {
  flex: 1,
};

export const $baseHeaderTitleStyle: ThemedStyle<HeaderTitleStyleProps> = (
  theme,
) => ({
  fontSize: 16,
  color: theme.colors.text,
});

export const $baseHeaderLargeTitleStyle: ThemedStyle<
  HeaderLargeTitleStyleProps
> = (theme) => ({
  color: theme.colors.text,
});

export const $baseContentStyle: ThemedStyle<ContentStyleProps> = (theme) => ({
  backgroundColor: theme.colors.background,
});

export const $baseHeaderStyle: ThemedStyle<HeaderStyleProps> = (theme) => ({
  backgroundColor: theme.colors.background,
});
