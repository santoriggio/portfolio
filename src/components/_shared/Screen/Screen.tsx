import { useAppTheme } from "@/hooks";
import { Stack } from "expo-router";
import { FC, PropsWithChildren, useEffect } from "react";
import {
  KeyboardAvoidingView,
  Platform,
  StyleProp,
  ViewStyle,
} from "react-native";
import { NativeStackNavigationOptions } from "@react-navigation/native-stack";
import {
  HeaderTitleStyleProps,
  HeaderStyleProps,
  ContentStyleProps,
  HeaderLargeTitleStyleProps,
  HeaderTintColorProps,
} from "./Screen.types";
import {
  $baseContentStyle,
  $basePageStyle,
  $baseHeaderLargeTitleStyle,
  $baseHeaderTitleStyle,
  $baseHeaderStyle,
} from "./Screen.styles";
import { TxKeyPath, translate } from "@/i18n";

export type ScreenProps = NativeStackNavigationOptions & {
  title?: string;
  tx?: TxKeyPath;
  style?: StyleProp<ViewStyle>;
  keyboardVerticalOffset?: number;
};

export const Screen: FC<PropsWithChildren<ScreenProps>> = ({
  title,
  tx,
  children,
  headerTitleStyle,
  headerLargeTitleStyle,
  contentStyle,
  headerTintColor,
  headerStyle,
  style,
  keyboardVerticalOffset,
  ...rest
}) => {
  useEffect(() => {
    if (title) {
      document.title = title;
    }

    if (tx) {
      document.title = translate(tx);
    }
  }, [title, tx]);
  const { themed, theme } = useAppTheme();

  const $pageStyle = [$basePageStyle, style];

  const $headerTitleStyle: HeaderTitleStyleProps = [
    themed($baseHeaderTitleStyle),
    headerTitleStyle,
  ];

  const $headerLargeTitleStyle: HeaderLargeTitleStyleProps = [
    themed($baseHeaderLargeTitleStyle),
    headerLargeTitleStyle,
  ];

  const $contentStyle: ContentStyleProps = [
    themed($baseContentStyle),
    contentStyle,
  ];

  const $headerTintColor: HeaderTintColorProps =
    headerTintColor || theme.colors.primary;

  const $headerStyle: HeaderStyleProps = [
    themed($baseHeaderStyle),
    headerStyle,
  ];

  return (
    <KeyboardAvoidingView
      style={$pageStyle}
      behavior="padding"
      keyboardVerticalOffset={keyboardVerticalOffset}
    >
      <Stack.Screen
        options={{
          title: tx ? translate(tx) : title,
          headerTitleStyle: $headerTitleStyle,
          headerLargeTitleStyle: $headerLargeTitleStyle,
          contentStyle: $contentStyle,
          headerTintColor: $headerTintColor,
          headerShadowVisible: false,
          headerLargeTitleShadowVisible: false,
          navigationBarColor: theme.colors.background,
          statusBarTranslucent: true,
          headerStyle: $headerStyle,
          ...rest,
        }}
      />
      {children}
    </KeyboardAvoidingView>
  );
};
