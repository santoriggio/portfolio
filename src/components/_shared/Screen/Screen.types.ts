import { NativeStackNavigationOptions } from "@react-navigation/native-stack";

type ExtractProp<P extends keyof NativeStackNavigationOptions> =
  NativeStackNavigationOptions[P];

export type HeaderTitleStyleProps = ExtractProp<"headerTitleStyle">;
export type HeaderLargeTitleStyleProps = ExtractProp<"headerLargeTitleStyle">;
export type ContentStyleProps = ExtractProp<"contentStyle">;
export type HeaderTintColorProps = ExtractProp<"headerTintColor">;
export type HeaderStyleProps = ExtractProp<"headerStyle">;
