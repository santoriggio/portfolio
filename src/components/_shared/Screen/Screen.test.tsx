import { Screen } from "./Screen";
import { renderRouter, screen } from "expo-router/testing-library";

describe("<Screen />", () => {
  it("renders correcly", () => {
    const MockComponent = jest.fn(() => <Screen />);

    renderRouter({
      index: MockComponent,
    });

    expect(screen).toMatchSnapshot();
  });
});
