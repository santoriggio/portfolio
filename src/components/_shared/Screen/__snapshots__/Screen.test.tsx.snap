// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Screen/> renders correcly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    />
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;
